Added processing of OperationalError raised when scheduling requests.

lipoja opened the pull request:
- created decorator which can be used for catching error
- failed state is set to all relevant/unprocessed requests
- sends messages with changes
- added unit test
- catching OperationalError during batch request processing

Refers to CLOUDDST-623 for more info.

midnightercz commented on tests/test_web/test_rabbitmq.py:
```diff
@@ -0,0 +1,115 @@
+
+@mock.patch('iib.web.api_v1.handle_regenerate_bundle_request')
+@mock.patch('iib.web.api_v1.messaging.send_message_for_state_change')
+def test_catch_regenerate_bundle_success(mock_smfsc, mock_hrbr, db, auth_env, client):
```
I don't get why you chose to name every test case as _success. Isn't point of test to test failure?

yashvardhannanavati commented on iib/web/api_v1.py:
```diff
@@ -438,21 +443,39 @@ def regenerate_bundle_batch():
+    build_and_requests = zip(payload['build_requests'], requests)
+    try:
+        for build_request, request in build_and_requests:
+
```
Could you please remove this extra line?

lipoja replied:
Yes

yashvardhannanavati commented on iib/web/api_v1.py:
```diff
@@ -438,21 +443,39 @@ def regenerate_bundle_batch():
+                continue
+
+            failed_ids.append(str(request.id))
+            request.add_state("failed", f"Scheduling of request failed.")
```
Could you please use single quotes to keep it consistent everywhere? 
I don't think f string is required here. `f'Scheduling of request failed.'` => `'Scheduling of request failed.'`

lipoja replied:
OK, thanks I did not notice that

yashvardhannanavati commented on iib/web/api_v1.py:
```diff
@@ -438,21 +443,39 @@ def regenerate_bundle_batch():
+            messaging.send_message_for_state_change(request)
+
+        flask.current_app.logger.exception(
+            f"Scheduling of request with ids {failed_ids} failed.\nException: {e}"
```
Same comment about using single quotes. We use single quotes in the entire code base. :)

yashvardhannanavati replied:
Also, shouldn't failed_ids be represented as  `', '.join(failed_ids)`

lipoja replied:
It should, fixed

yashvardhannanavati commented on iib/web/api_v1.py:
```diff
@@ -438,21 +443,39 @@ def regenerate_bundle_batch():
+            f"Scheduling of request with ids {failed_ids} failed.\nException: {e}"
         )
 
+        raise IIBError(f"Scheduling of requests with ids {failed_ids} failed.")
```
Same as above.

yashvardhannanavati commented on iib/web/exception_handlers.py:
```diff
@@ -0,0 +1,32 @@
+
+
+def catch_rabbit(request, messaging):
+    """Catches RabbitMQ error and raises IIBError instead."""
```
Could you please update the doc string to reflect the parameters taken by this function?

lipoja replied:
Yes, updated

lipoja commented on iib/web/exception_handlers.py:
```diff
@@ -0,0 +1,32 @@
+from iib.exceptions import IIBError
+
+
+def catch_rabbit(request, messaging):
```
Yes, that is batter way, thanks.

yashvardhannanavati commented on iib/web/exception_handlers.py:
```diff
@@ -0,0 +1,37 @@
+from iib.web import messaging
+
+
+def catch_rabbit(request):
```
Can we rename this to `catch_rabbitmq_error` or something better please?

lipoja replied:
sure, no more catching rabbits it is ;)

mprahl commented on iib/web/exception_handlers.py:
```diff
@@ -0,0 +1,37 @@
+from iib.web import messaging
+
+
+def catch_rabbitmq_error(request):
```
Could you please make the name more generic? The fact that IIB uses RabbitMQ as its broker is an implementation detail of the deployment and not the code.

A name such as `catch_broker_error` would likely be more appropriate.

mprahl commented on iib/web/api_v1.py:
```diff
@@ -439,21 +444,38 @@ def regenerate_bundle_batch():
+            request_id_strs.append(str(request.id))
+    except OperationalError as e:
+        failed_ids = []
+        for _, request in build_and_requests:
```
Could you iterate through `requests` instead of `build_and_requests` since you don't use the first element in the tuple?

mprahl commented on iib/web/api_v1.py:
```diff
@@ -439,21 +444,38 @@ def regenerate_bundle_batch():
+            messaging.send_message_for_state_change(request)
+
+        flask.current_app.logger.exception(
+            f'Scheduling of request with ids {", ".join(failed_ids)} failed.\nException: {e}'
```
It's unnecessary to log the exception message in the log message itself since you are doing `logger.exception`, which will log the exception and traceback.
```suggestion
            f'Scheduling of request with IDs {", ".join(failed_ids)} failed'
```

mprahl commented on iib/web/api_v1.py:
```diff
@@ -439,21 +444,38 @@ def regenerate_bundle_batch():
+                continue
+
+            failed_ids.append(str(request.id))
+            request.add_state('failed', f'Scheduling of request failed.')
```
I prefer full sentences for user-facing messages. Also, this doesn't need to be an f-string.
```suggestion
            request.add_state('failed', 'The scheduling of the request failed')
```

mprahl commented on iib/web/api_v1.py:
```diff
@@ -439,21 +444,38 @@ def regenerate_bundle_batch():
     # This list will be used for the log message below and avoids the need of having to iterate
     # through the list of requests another time
     request_id_strs = []
```
Now that this being used for things other than log messages, it might be better to rename this variable and not have it contain strings:
```suggestion
    processed_request_ids = []
```

mprahl commented on iib/web/api_v1.py:
```diff
@@ -439,21 +444,38 @@ def regenerate_bundle_batch():
+            f'Scheduling of request with ids {", ".join(failed_ids)} failed.\nException: {e}'
         )
 
+        raise IIBError(f'Scheduling of requests with ids {", ".join(failed_ids)} failed.')
```
```suggestion
        raise IIBError(f'The scheduling of the build requests with IDs {", ".join(failed_ids)} failed.')
```

mprahl commented on iib/web/exception_handlers.py:
```diff
@@ -0,0 +1,37 @@
+            try:
+                return func(*args, **kwargs)
+            except OperationalError as e:
+
```
```suggestion
```

mprahl commented on iib/web/exception_handlers.py:
```diff
@@ -0,0 +1,37 @@
+
+                payload = flask.request.get_json()
+                req = request.from_json(payload)
+                req.add_state("failed", f"Scheduling of request failed.")
```
This doesn't need to be an f-string and I prefer full sentences for user-facing messages.
```suggestion
                req.add_state("failed", f"The scheduling of the build request failed")
```

mprahl commented on iib/web/exception_handlers.py:
```diff
@@ -0,0 +1,37 @@
+                messaging.send_message_for_state_change(request)
+
+                flask.current_app.logger.exception(
+                    f"Scheduling of request id {req.id} failed.\nException: {e}"
```
Same comment as above:
```suggestion
                    f"The scheduling of the build request with ID {req.id} failed"
```

mprahl commented on iib/web/exception_handlers.py:
```diff
@@ -0,0 +1,37 @@
+                    f"Scheduling of request id {req.id} failed.\nException: {e}"
+                )
+
+                raise IIBError(f"Scheduling of request id {req.id} failed.")
```
```suggestion
                raise IIBError(f"The scheduling of the build request with the ID {req.id} failed")
```

mprahl commented on tests/test_web/test_rabbitmq.py:
```diff
@@ -0,0 +1,115 @@
+def assert_testing(rv, mock_smfsc):
+    response_json = {'error': "Scheduling of request id 2 failed."}
+    assert rv.status_code == 500
+    assert response_json == rv.json
```
It's less confusing if the same side is always picked for the actual value and the expected value
```suggestion
    assert rv.json == response_json
```

mprahl commented on tests/test_web/test_rabbitmq.py:
```diff
@@ -0,0 +1,115 @@
+    assert_testing(rv, mock_smfsc)
+
+
+@pytest.mark.parametrize(
```
This parametrize doesn't add any value. The tests it performs are well covered elsewhere.

mprahl commented on iib/web/exception_handlers.py:
```diff
@@ -0,0 +1,37 @@
+import functools
```
Could move this code to `iib/web/errors.py`? That Python module seems like a good fit for this.

mprahl commented on tests/test_web/test_rabbitmq.py:
```diff
@@ -0,0 +1,115 @@
+@mock.patch('iib.web.api_v1.handle_regenerate_bundle_request')
+@mock.patch('iib.web.api_v1.messaging.send_message_for_state_change')
+@mock.patch('iib.web.api_v1.messaging.send_messages_for_new_batch_of_requests')
+def test_catch_regenerate_bundle_batch_failure(
```
Bonus points if you query the database to ensure that the state of the requests 2 and 3 are set to `failed`.

mprahl commented on tests/test_web/test_rabbitmq.py:
```diff
@@ -0,0 +1,115 @@
+
+@mock.patch('iib.web.api_v1.handle_rm_request')
+@mock.patch('iib.web.api_v1.messaging.send_message_for_state_change')
+def test_catch_remove_operator_failure(mock_smfsc, mock_rm, db, auth_env, client):
```
Bonus points if you query the database to see if the request is in the failed state.

mprahl commented on tests/test_web/test_rabbitmq.py:
```diff
@@ -0,0 +1,115 @@
+
+@mock.patch('iib.web.api_v1.handle_regenerate_bundle_request')
+@mock.patch('iib.web.api_v1.messaging.send_message_for_state_change')
+def test_catch_regenerate_bundle_failure(mock_smfsc, mock_hrbr, db, auth_env, client):
```
Bonus points if you query the database to see if the request is in the failed state.

mprahl commented on tests/test_web/test_rabbitmq.py:
```diff
@@ -0,0 +1,115 @@
+
+@mock.patch('iib.web.api_v1.handle_add_request')
+@mock.patch('iib.web.api_v1.messaging.send_message_for_state_change')
+def test_catch_add_bundle_failure(mock_smfsc, mock_har, db, auth_env, client):
```
Bonus points if you query the database to see if the request is in the failed state.

yashvardhannanavati commented:
@lipoja Could you please address @mprahl's comments?

lipoja commented:
> @lipoja Could you please address @mprahl's comments?

Sure, I will do it as a first thing when I am back from PTO. I've already read through those comments and the changes should be  done quickly. Thanks for those comments.

yashvardhannanavati commented:
No rush, Enjoy your holidays! :)

lipoja commented:
@mprahl Hi, could you have a look on latest changes? I've updated the code based on your comments. I also added inserting faild state to database. It looks like previous version did not change the state in DB at all.
Thanks.

mprahl commented:
@lipoja because #120 was merged, the scope of this PR needs to be extended unfortunately :frowning_face:

mprahl commented on iib/web/api_v1.py:
```diff
@@ -438,25 +443,46 @@ def regenerate_bundle_batch():
+                continue
+
+            failed_ids.append(str(request.id))
+            request.add_state('failed', 'The scheduling of the build request failed.')
```
The rest of the state reasons in the code do not have periods at the end, so let's remove that to be consistent.
```suggestion
            request.add_state('failed', 'The scheduling of the build request failed')
```

mprahl commented on iib/web/api_v1.py:
```diff
@@ -438,25 +443,46 @@ def regenerate_bundle_batch():
+
+        db.session.commit()
+        flask.current_app.logger.exception(
+            f'The scheduling of the build requests with IDs {", ".join(failed_ids)} failed.'
```
```suggestion
            f'The scheduling of the build requests with IDs {", ".join(failed_ids)} failed'
```

mprahl commented on iib/web/api_v1.py:
```diff
@@ -438,25 +443,46 @@ def regenerate_bundle_batch():
+        )
+
+        raise IIBError(
+            f'The scheduling of the build requests with IDs {", ".join(failed_ids)} failed.'
```
```suggestion
            f'The scheduling of the build requests with IDs {", ".join(failed_ids)} failed'
```

mprahl commented on iib/web/errors.py:
```diff
@@ -32,3 +35,34 @@ def json_error(error):
+            except kombu.exceptions.OperationalError:
+                payload = request.get_json()
+                req = iib_request.from_json(payload)
+                req.add_state("failed", "The scheduling of the build request failed.")
```
```suggestion
                req.add_state("failed", "The scheduling of the build request failed")
```

mprahl commented on iib/web/errors.py:
```diff
@@ -32,3 +35,34 @@ def json_error(error):
+                messaging.send_message_for_state_change(iib_request)
+
+                current_app.logger.exception(
+                    f"The scheduling of the build request with ID {req.id} failed."
```
```suggestion
                    f"The scheduling of the build request with ID {req.id} failed"
```

mprahl commented on iib/web/errors.py:
```diff
@@ -32,3 +35,34 @@ def json_error(error):
+                    f"The scheduling of the build request with ID {req.id} failed."
+                )
+
+                raise IIBError(f"The scheduling of the build request with ID {req.id} failed.")
```
```suggestion
                raise IIBError(f"The scheduling of the build request with ID {req.id} failed")
```

mprahl commented on tests/test_web/test_broker_error.py:
```diff
@@ -0,0 +1,116 @@
+    assert response_json == rv.json
+    assert mock_smfsc.call_count == 2
+
+    req_state = db.session.query(Request).filter(Request.id == 2).one()
```
```suggestion
    req_state = db.session.query(Request).get(2)
```

mprahl replied:
This needs to be getting the request with the ID of `1`. Since your test starts with a clean database, there should be only the request that you created as part of the test.

mprahl commented on iib/web/errors.py:
```diff
@@ -32,3 +35,34 @@ def json_error(error):
+                return func(*args, **kwargs)
+            except kombu.exceptions.OperationalError:
+                payload = request.get_json()
+                req = iib_request.from_json(payload)
```
This actually creates a new `Request` object, so you'll end up with duplicate requests. One will still be `in_progress` and the other will be `failed`. You need to filter for the request that was just created and failed in this request.

mprahl commented on iib/web/api_v1.py:
```diff
@@ -438,25 +443,46 @@ def regenerate_bundle_batch():
+
+            failed_ids.append(str(request.id))
+            request.add_state('failed', 'The scheduling of the build request failed.')
+            db.session.add(request)
```
Optional: I think this can be removed

lipoja commented:
> @lipoja because #120 was merged, the scope of this PR needs to be extended unfortunately

I know, I am already working on covering that. Thanks.

lipoja commented:
@yashvardhannanavati @mprahl PTAL 

I had to change a way of implementing this issue. From what I know I am not able to use decorator for most of the tasks, because a request is added in database inside a decorated function. So I do not know correct ID of the row in DB. And I guess there is possibility to have two or more requests with same parameters, so searching just by these parameters in DB could not get me the correct result. Therefore I've modified the code to have just functions which deals with setting requests in DB as failed and sending messages. 
Please have a look on it, if the approach is wrong then write me some comments. I am open to changes. 
Thanks.

mprahl commented on iib/web/errors.py:
```diff
@@ -32,3 +34,45 @@ def json_error(error):
+
+def handle_broker_error(request_id, request_model):
+    """
+    Handle RabbitMQ error by sating request as failed and raises IIBError.
```
```suggestion
    Handle broker errors by setting the request as failed and raise an IIBError exception.
```

mprahl commented on iib/web/errors.py:
```diff
@@ -32,3 +34,45 @@ def json_error(error):
     return response
+
+
+def handle_broker_error(request_id, request_model):
```
Could you please just accept the request object so that you don't have to query the database for it?

mprahl commented on iib/web/errors.py:
```diff
@@ -32,3 +34,45 @@ def json_error(error):
+    db.session.commit()
+    messaging.send_message_for_state_change(req)
+
+    current_app.logger.exception(f"The scheduling of the build request with ID {req.id} failed")
```
Please assign this string to a variable so you don't have to duplicate it and do the string substitution twice.

mprahl commented on iib/web/api_v1.py:
```diff
@@ -438,26 +451,30 @@ def regenerate_bundle_batch():
+            request_jsons.append(request.to_json())
+            processed_request_ids.append(str(request.id))
+    except OperationalError:
+        handle_broker_batch_error(requests, processed_request_ids)
```
Could you just pass in the request objects to fail directly instead?
```suggestion
        handle_broker_batch_error(requests[len(processed_request_ids):])
```

mprahl commented on iib/web/errors.py:
```diff
@@ -32,3 +34,45 @@ def json_error(error):
+            continue
+
+        failed_ids.append(str(req.id))
+        req.add_state('failed', 'The scheduling of request failed')
```
```suggestion
        req.add_state('failed', 'The scheduling of the request failed')
```

mprahl commented on iib/web/errors.py:
```diff
@@ -32,3 +34,45 @@ def json_error(error):
+
+    db.session.commit()
+    current_app.logger.exception(
+        f'The scheduling of the build requests with IDs {", ".join(failed_ids)} failed'
```
Could you please assign this string to a variable to avoid code duplication and doing the string substitution twice?

yashvardhannanavati commented on iib/web/errors.py:
```diff
@@ -32,3 +34,38 @@ def json_error(error):
+
+    :param Request request: Request which will be set as failed
+    """
+    request.add_state("failed", "The scheduling of the request failed")
```
Can we use single quotes throughout the change please.

yashvardhannanavati commented on iib/web/errors.py:
```diff
@@ -32,3 +34,38 @@ def json_error(error):
+    """
+    Handle broker errors by setting the request as failed and raise an IIBError exception.
+
+    :param Request request: Request which will be set as failed
```
Could we please add the :raises IIBError... in the docstring please?

yashvardhannanavati commented on iib/web/errors.py:
```diff
@@ -32,3 +34,38 @@ def json_error(error):
+    """
+    Handle broker errors by setting all request as failed and raise an IIBError exception.
+
+    :param list requests: list of all requests that should be marked as failed
```
Could we please add the :raises IIBError... in the docstring ?

lipoja replied:
Thanks, done.

yashvardhannanavati commented on tests/test_web/test_broker_error.py:
```diff
@@ -0,0 +1,218 @@
+
+
+def assert_testing(rv, mock_smfsc, db):
+    response_json = {'error': "The scheduling of the build request with ID 1 failed"}
```
Could you use single quotes please.

lipoja replied:
I guess PyCharm is set to modify it automatically, so I did not notice that at first. But when you pointed out it looks like all recently edited strings were transformed to double quote. 
Fixed.

mprahl commented on iib/web/errors.py:
```diff
@@ -1,9 +1,11 @@
 # SPDX-License-Identifier: GPL-3.0-or-later
-from flask import jsonify
+
```
```suggestion
```

mprahl commented on iib/web/errors.py:
```diff
@@ -32,3 +34,40 @@ def json_error(error):
+
+def handle_broker_batch_error(requests):
+    """
+    Handle broker errors by setting all request as failed and raise an IIBError exception.
```
```suggestion
    Handle broker errors by setting all requests as failed and raise an IIBError exception.
```

mprahl commented on tests/test_web/test_broker_error.py:
```diff
@@ -0,0 +1,218 @@
+    assert response_json == rv.json
+    assert mock_smfsc.call_count == 2
+
+    req_state = db.session.query(Request).filter(Request.id == 1).one()
```
```suggestion
    req_state = db.session.query(Request).get(1)
```

mprahl commented on tests/test_web/test_broker_error.py:
```diff
@@ -0,0 +1,218 @@
+    mock_smfnbor, mock_smfsc, mock_hrbr, app, auth_env, client, db,
+):
+
+    user_to_queue = {'tbrady@DOMAIN.LOCAL': 'Patriots'}
+    expected_queue = 'Patriots'
+    annotations = None
+
+    mock_hrbr.apply_async.side_effect = OperationalError
+    app.config['IIB_USER_TO_QUEUE'] = user_to_queue
```
This is all sufficiently tested elsewhere. I'd like to avoid duplicating tests. Could you please remove this part of the test?

mprahl commented on tests/test_web/test_broker_error.py:
```diff
@@ -0,0 +1,218 @@
+            {'from_bundle_image': 'registry.example.com/bundle-image3:latest'},
+        ]
+    }
+    if annotations:
```
This is always `None` right? Just delete this line and the next.

mprahl commented on tests/test_web/test_broker_error.py:
```diff
@@ -0,0 +1,218 @@
+    assert rv.json == response_json
+    assert mock_smfsc.call_count == 3
+    assert mock_hrbr.apply_async.call_count == 1
+    mock_hrbr.apply_async.assert_has_calls(
+        (
+            mock.call(
+                args=['registry.example.com/bundle-image:latest', None, 1],
+                link_error=mock.ANY,
+                queue=expected_queue,
+            ),
+        )
+    )
```
Related to my previous comment, this check is sufficiently covered in another test.

mprahl commented on tests/test_web/test_broker_error.py:
```diff
@@ -0,0 +1,218 @@
+    req_states = (
+        db.session.query(RequestState)
+        .join(Request, Request.request_state_id == RequestState.id)
+        .filter(Request.id.in_((1, 2, 3)))
```
```suggestion
        .filter(Request.id.in_((1, 2)))
```

mprahl commented on tests/test_web/test_broker_error.py:
```diff
@@ -0,0 +1,218 @@
+    req_add = db.session.query(RequestAdd).get(1)
+    req_rm = db.session.query(RequestRm).get(2)
+    # First request is processed because we are testing failing on RequestRM
+    assert req_add.state.state == RequestStateMapping.in_progress.value
```
This approach is cleaner than doing an additional query in `test_add_rm_batch_add_failure`. Could you please adjust it in `test_add_rm_batch_add_failure`?

mprahl commented on iib/web/api_v1.py:
```diff
@@ -438,26 +451,31 @@ def regenerate_bundle_batch():
-        )
+            request_jsons.append(request.to_json())
+            processed_request_ids.append(str(request.id))
+    except OperationalError:
```
Can you be consistent and either always use `OperationalError` or `kombu.exceptions.OperationalError`?

mprahl commented on iib/web/api_v1.py:
```diff
@@ -542,18 +559,30 @@ def add_rm_batch():
+                    argsrepr=repr(safe_args),
+                    queue=celery_queue,
+                )
+        except OperationalError:
```
Can you be consistent and either always use `OperationalError` or `kombu.exceptions.OperationalError`?

lipoja replied:
Thanks, changed to kombu.exceptions.OperationalError
